src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs

```diff
@@ -3,311 +3,819 @@
+                    {
+                        continue;
+                    }
+
                     if (projectReferencePaths == null)
                     {
-                        projectReferencePaths = GetProjectReferencePaths(lockFile);
+                        projectReferencePaths = GetProjectReferencePaths(_lockFile);
                     }
 
                     if (!directProjectDependencies.Contains(library.Name))
                     {
-                        items.Add(new TaskItem(projectReferencePaths[library.Name]));
+                        WriteItem(projectReferencePaths[library.Name], library);
                     }
                 }
             }
 
-            return items.ToArray();
-        }
-
-        private ITaskItem[] RaiseAnalyzers(LockFile lockFile, LockFileTarget target)
-        {
-            var items = new List<ITaskItem>();
-            Dictionary<string, LockFileTargetLibrary> targetLibraries = null;
-
-            foreach (var library in lockFile.Libraries)
+            private void WriteItems<T>(
+                LockFileTarget target,
+                Func<LockFileTargetLibrary, IList<T>> getAssets,
+                Func<T, bool> filter = null,
+                Action<T> writeMetadata = null)
+                where T : LockFileItem
             {
-                if (library.IsPackage())
+                foreach (var library in target.Libraries)
                 {
-                    foreach (var file in library.Files)
+                    if (!library.IsPackage())
+                    {
+                        continue;
+                    }
+
+                    foreach (T asset in getAssets(library))
                     {
-                        if (NuGetUtils.IsApplicableAnalyzer(file, ProjectLanguage))
+                        if (asset.IsPlaceholderFile() || (filter != null && !filter.Invoke(asset)))
                         {
-                            if (targetLibraries == null)
-                            {
-                                targetLibraries = target.Libraries.ToDictionary(l => l.Name, StringComparer.OrdinalIgnoreCase);
-                            }
-
-                            if (targetLibraries.TryGetValue(library.Name, out var targetLibrary))
-                            {
-                                items.Add(CreatePackageAssetItem(targetLibrary, ResolvePackageAssetPath(targetLibrary, file)));
-                            }
+                            continue;
                         }
+
+                        string itemSpec = ResolvePackageAssetPath(library, asset.Path);
+                        WriteItem(itemSpec, library);
+                        writeMetadata?.Invoke(asset);
                     }
                 }
             }
 
-            return items.ToArray();
-        }
+            private void WriteItem(string itemSpec)
+            {
+                FlushMetadata();
```

Comment:
It seems like it would be quite easy to end up writing the metadata out of order, since metadata is flushed before an item is written instead of afterwards.  What's the reason for buffering the metadata items anyway?  Would it be more straightforward to have `WriteMetadata` write directly to the stream, and in `WriteItems` call the `writeMetadata` callback after the call to `WriteItem`?

Reply:
The reason for buffering is because we don't know how many metadata items will be written and we need to prepend the count. We have the same problem for items but there we write a placeholder and seek back to set the count. Seeking costs flushing the file stream buffer, though so we don't want to do that for every item. I was trying hard to avoid allocations in the writer so as not to over-penalize the first, uncached build. I will give some thought to how to improve this, but I won't hold on this for the initial PR.

Reply:
As discussed offline another approach to this would have been to make the first run go from assets.json to ITaskItems directly (as before), but also write out the cache file for the incremental builds and incremental builds alone to use. The main reason why I did not write it that way was to make it so that full builds and incremental builds get the items from the same code so that there won't be mysterious bugs that only show up on incremental builds. This means that the entire test suite is stressing the cache reading/writing now.

There is potential to improve the readability of the code and the performance of the first build here, but I am reluctant to tackle that and reset the preview2 measurements that have been taken with this approach, nor do I want to lose the coverage property.

Filed #2048 to think about changing the approach.